### ISSUE TITLE
Entity pop-in

### ISSUE THREAD

**Issue by Veyrdite** [Unvanquished/Unvanquished#309]
Players and structures 'pop' into existence as you go around corners.  You can't see this, but it causes other problems:

(1) Blips suddenly appear on the 'alien sense' at a close range, rather than gradually appearing from the maximum sense range as they should.
(2) Rather than gradually warning you of nearby structures as you get closer to them, sudden 'walls of sound' whack into you as multiple structures with idle sound effects simultaneously come into existence around a corner.

**Comment by bmorel** [comment]
I have a proposal for this.

The game already supports cluster representations. Maybe we could add clusters a "weight" which would impact from how far the cluster should *start* to send informations.

For example, image a cluster A of 1 turret, and a 2nd cluster B with 2 turrets, a telenode, an armoury and a medistation.
The B cluster would send *some* informations about itself sooner, starting with info of a single building, and more as the distance reduces.
Cluster A info would only be sent where cluster B would "only" disclose, say, 80% info about itself.

I think this solution could be an interesting middle-ground between cheat-easyness and sudden pop of a wall of sound.
This could also maybe impact alien sense or human radars as well: the more enemy stuff around, the further one can detect it, but also the more blurry it would be.

This could even open opportunities for new buildings that have a higher "noise score", or inversely that would reduce "noise score" of buildings nearby.

**Comment by Viech** [comment]
Assuming that you refer only to sounds, I do not see how base clusters would work around the following issues:

> We already tested an implementation where entities were sent as soon as they entered the alien sense/radar range, but this worked poorly together with sounds as you started hearing things from different parts of the map, no matter how many walls were between you and the source. We also tried sending entities that were positioned around up to two line-of-sight breaking corners (instead of just one) with similar results.

Base clusters are _probably_ not a good indication of map acoustics.

If you are talking about all entities and not just sounds, keep in mind that the primary concern is that any entity sent to the player will also be drawn if it is in front of that player, regardless of whether there is a wall in between. Entities are hidden (through the `vis` function) primarily for performance reason and some maps are even built in a way to facilitate this (e.g., the main entrances to ATCS/plat23 bases are designed as _vis blockers_).

Maybe a straightforward solution to part (2) of the original issue would be to just smoothen the set-in and set-out of immobile sounds coming from buildables and map speakers.

**Comment by slipher** [comment]
If the potentially visible set is used to decide whether we might be able to see enemies, then clearly a potentially audible set is what we need to know whether we might be able to hear them :stuck_out_tongue:  Of course, that wouldn't be enough by itself since you'd have to dampen based on walls blocking, not just straight-line distance.


> One visible consequence of this, though, is that sometimes the creep of an invisible (around the corner) entity isn't visible, as the client is not aware of the structure.

This seems relatively easily solvable by using a larger bounding box accommodating all visual effects for the alien structure entity in visibility calculations.

**Comment by illwieckz** [comment]
> keep in mind that the primary concern is that any entity sent to the player will also be drawn if it is in front of that player, regardless of whether there is a wall in between. Entities are hidden (through the `vis` function) primarily for performance reason

I guess another primary concern is anticheat. One can write a software that would draw all entities above everything else (wallhack). Though with our beacon systems it's almost like we already have a wallhack (we may want to reserve precise beacons only for a couple of things, but that's another topic). There also exists some wallhack in the field drawing the entity location from sound position…

**Comment by Viech** [comment]
> I guess another primary concern is anticheat.

Indeed, I also mentioned this earlier (aka many years ago) in this thread, though I feel that performance must have been the root idea behind `vis`, coming from the era of LAN parties. Performance certainly is the reason why some maps have vis-blockers in them. There are also explicit mapping techniques to obtain performance improvements through tighter vis (`hint` and `portal` shaders).

So while we could send dedicated _sound entities_ earlier, buildable entities that _also make a sound_ should be held back independent of cheating consideration to not degrade performance.
